File: src/openinsure/domain/mga.py
"""MGA/Delegated Authority domain entities."""

from datetime import date
from decimal import Decimal

from pydantic import Field

from openinsure.domain.common import DomainEntity, Money


class DelegatedAuthority(DomainEntity):
    """Represents an MGA's delegated authority granted by the carrier."""

    mga_name: str
    mga_id: str
    status: str = "active"  # active, suspended, expired
    effective_date: date | None = None
    expiration_date: date | None = None
    lines_of_business: list[str] = Field(default_factory=list)
    premium_authority: Money = Decimal("0")
    premium_written: Money = Decimal("0")
    claims_authority: Money = Decimal("0")
    loss_ratio: Decimal = Decimal("0")
    compliance_score: Decimal = Decimal("100")
    last_audit_date: date | None = None


class BordereauRecord(DomainEntity):
    """A bordereau submission from an MGA for a given period."""

    mga_id: str
    period: str  # e.g., "2026-Q1"
    premium_reported: Money = Decimal("0")
    claims_reported: Money = Decimal("0")
    loss_ratio: Decimal = Decimal("0")
    policy_count: int = 0
    claim_count: int = 0
    status: str = "pending"  # pending, validated, exceptions, accepted
    exceptions: list[str] = Field(default_factory=list)

File: dashboard/src/api/mga.ts
import client from './client';

export interface MGAAuthority {
  mga_id: string;
  mga_name: string;
  status: string;
  effective_date: string | null;
  expiration_date: string | null;
  lines_of_business: string[];
  premium_authority: number;
  premium_written: number;
  claims_authority: number;
  loss_ratio: number;
  compliance_score: number;
  last_audit_date: string | null;
}

export interface MGABordereau {
  id: string;
  mga_id: string;
  period: string;
  premium_reported: number;
  claims_reported: number;
  loss_ratio: number;
  policy_count: number;
  claim_count: number;
  status: string;
  exceptions: string[];
}

export interface MGAPerformance {
  total_mgas: number;
  active_mgas: number;
  suspended_mgas: number;
  total_premium_written: number;
  total_premium_authority: number;
  average_loss_ratio: number;
  average_compliance_score: number;
  authorities: MGAAuthority[];
}

export async function getMGAAuthorities(status?: string): Promise<MGAAuthority[]> {
  const params = status ? { status } : {};
  const { data } = await client.get<MGAAuthority[]>('/mga/authorities', { params });
  return data;
}

export async function getMGAAuthority(mgaId: string): Promise<MGAAuthority> {
  const { data } = await client.get<MGAAuthority>(`/mga/authorities/${mgaId}`);
  return data;
}

export async function getMGABordereaux(mgaId?: string): Promise<MGABordereau[]> {
  const params = mgaId ? { mga_id: mgaId } : {};
  const { data } = await client.get<MGABordereau[]>('/mga/bordereaux', { params });
  return data;
}

export async function getMGAPerformance(): Promise<MGAPerformance> {
  const { data } = await client.get<MGAPerformance>('/mga/performance');
  return data;
}
